nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+  std::stack<LayerNode> Stack;
+
+  bool *visited = new bool[num_node];
+  for (unsigned int i = 0; i < num_node; ++i)
+    visited[i] = false;
+
+  for (unsigned int i = 0; i < num_node; ++i) {
+    if (visited[i] == false) {
+      topologicalSortUtil(i, visited, Stack);
+    }
+  }
+
+  delete[] visited;
+
+  while (Stack.empty() == false) {
+    Sorted.push_back(Stack.top());
+    Stack.pop();
+  }
+
+  for (unsigned int i = 0; i < Sorted.size(); ++i) {
+    std::cout << Sorted[i].layer->getName() << " ";
+  }
+  std::cout << std::endl;
+}
+
+void NetworkGraph::ensureName(std::shared_ptr<Layer> layer,
+                              const std::string &prefix = "",
+                              bool force_rename = false) {
+  std::string orig_name = layer->getName();
+  bool orig_name_empty = orig_name.empty();
+  if (!orig_name_empty && !force_rename &&
+      layer_names.end() == layer_names.find(orig_name))
+    return;
+
+  /** If just prefix with layer name makes it unique - directly set the name */
+  if (!orig_name_empty) {
+    std::string direct_name = prefix + orig_name;
+    if (layer_names.find(direct_name) == layer_names.end()) {
+      layer->setName(direct_name);
+      return;
+    }
+  }
+
+  std::set<std::string>::iterator iter;
+  std::string name;
+  if (orig_name_empty)
+    orig_name = layer->getType();
+  std::string direct_name = prefix + orig_name;
+
+  do {
+    name = direct_name + std::to_string(def_name_count++);
+    iter = layer_names.find(name);
+  } while (iter != layer_names.end());
+
+  layer->setName(name);
+}
+
+int NetworkGraph::realizeMultiInputType(Layer &current) {
+  int status = ML_ERROR_NONE;
+  if (current.num_inputs == 1)
+    return ML_ERROR_NONE;
+
+  if (current.num_inputs > 1) {
+    std::shared_ptr<Layer> layer = nntrainer::createLayer("concat");
+    ensureName(layer, current.getName());
+    layer->num_inputs = current.num_inputs;
+    layer->input_layers.clear();
+    for (unsigned int i = 0; i < current.input_layers.size(); ++i)
+      layer->input_layers.push_back(current.input_layers[i]);
+
+    current.num_inputs = 1;
+    current.input_layers.clear();
+    current.input_layers.push_back(layer->getName());
+    addLayerNode(layer);
+  }
+  return status;
+}
+
+int NetworkGraph::realizeFlattenType(
+  Layer &current, std::vector<std::shared_ptr<Layer>> layers) {
+  if (num_node == 0) {
+    ml_loge("layer is empty");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (current.getType() == "flatten") {
+    ml_loge(
+      "It is not allowed to realize flatten layer, possibly flatten layer is "
+      "added right after flatten");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  std::shared_ptr<Layer> layer = nntrainer::createLayer("flatten");
+
+  ensureName(layer, current.getName());
+
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(current.getName());
+
+  addLayerNode(layer);
+
+  for (unsigned int i = 0; i < layers.size(); ++i) {
```

Comment:
please create a function like `updateNameInLayers(std::vector<std::shared_ptr<Layer>> layers, const std::string &name)` which does this. This is currently repeatedly written in many places.

Reply:
Yes. I will. 
> Also move this call out of here to the original function. This function should just realize flatten layer. The handling of side-effects should take place in the caller function.

I don't think it it be out of here. This is only for adding flatten layer and If it is not possible, then it is the error of adding flatten layer. ( If it has error during this addLayerNode, then it shouldn't go further I guess )

Reply:
As any error in creating graph should not change the graph state, or leave graph in un-defined state, giving error before or after does not change. The behavior of updating name is associated with the caller - where a subgraph is to be added - but has nothing to do with realization of the layer.

Reply:
> As any error in creating graph should not change the graph state, or leave graph in un-defined state, giving error before or after does not change. The behavior of updating name is associated with the caller - where a subgraph is to be added - but has nothing to do with realization of the layer.

For this PR, realization means adding Layer which associated with input and output layers setting. we make split the function with realization, adding layer with setting input and output layer then. But this can be done later PR.

Reply:
ok.